src/node_api.cc

```diff
@@ -0,0 +1,2430 @@
+  return reinterpret_cast<HandleScopeWrapper*>(s);
+}
+
+napi_escapable_handle_scope JsEscapableHandleScopeFromV8EscapableHandleScope(
+    EscapableHandleScopeWrapper* s) {
+  return reinterpret_cast<napi_escapable_handle_scope>(s);
+}
+
+EscapableHandleScopeWrapper*
+V8EscapableHandleScopeFromJsEscapableHandleScope(
+    napi_escapable_handle_scope s) {
+  return reinterpret_cast<EscapableHandleScopeWrapper*>(s);
+}
+
+//=== Conversion between V8 Handles and napi_value ========================
+
+// This is assuming v8::Local<> will always be implemented with a single
+// pointer field so that we can pass it around as a void* (maybe we should
+// use intptr_t instead of void*)
+
+napi_value JsValueFromV8LocalValue(v8::Local<v8::Value> local) {
+  // This is awkward, better way? memcpy but don't want that dependency?
+  union U {
+    napi_value v;
+    v8::Local<v8::Value> l;
+    U(v8::Local<v8::Value> _l) : l(_l) {}
+  } u(local);
+  assert(sizeof(u.v) == sizeof(u.l));
+  return u.v;
+}
+
+v8::Local<v8::Value> V8LocalValueFromJsValue(napi_value v) {
+  // Likewise awkward
+  union U {
+    napi_value v;
+    v8::Local<v8::Value> l;
+    U(napi_value _v) : v(_v) {}
+  } u(v);
+  assert(sizeof(u.v) == sizeof(u.l));
+  return u.l;
+}
+
+static v8::Local<v8::Function> V8LocalFunctionFromJsValue(napi_value v) {
+  // Likewise awkward
+  union U {
+    napi_value v;
+    v8::Local<v8::Function> f;
+    U(napi_value _v) : v(_v) {}
+  } u(v);
+  assert(sizeof(u.v) == sizeof(u.f));
+  return u.f;
+}
+
+// Wrapper around v8::Persistent that implements reference counting.
+class Reference {
+ public:
+  Reference(v8::Isolate* isolate,
+            v8::Local<v8::Value> value,
+            int initialRefcount,
+            bool deleteSelf,
+            napi_finalize finalizeCallback = nullptr,
+            void* finalizeData = nullptr,
+            void* finalizeHint = nullptr)
+       : _isolate(isolate),
+        _persistent(isolate, value),
+        _refcount(initialRefcount),
+        _deleteSelf(deleteSelf),
+        _finalizeCallback(finalizeCallback),
+        _finalizeData(finalizeData),
+        _finalizeHint(finalizeHint) {
+    if (initialRefcount == 0) {
+      _persistent.SetWeak(
+          this, FinalizeCallback, v8::WeakCallbackType::kParameter);
+      _persistent.MarkIndependent();
+    }
+  }
+
+  ~Reference() {
+    // The V8 Persistent class currently does not reset in its destructor:
+    // see NonCopyablePersistentTraits::kResetInDestructor = false.
+    // (Comments there claim that might change in the future.)
+    // To avoid memory leaks, it is better to reset at this time, however
+    // care must be taken to avoid attempting this after the Isolate has
+    // shut down, for example via a static (atexit) destructor.
+    _persistent.Reset();
+  }
+
+  int AddRef() {
+    if (++_refcount == 1) {
+      _persistent.ClearWeak();
+    }
+
+    return _refcount;
+  }
+
+  int Release() {
```

Comment:
@jasongin Exactly, the other uses in Node are also somewhat internal-y. And also, I agree, it’s obviously good to have an engine-agnostic API – I don’t think V8 actually exposes anything like this anyway, though. :)